Review of the pageto adapt it into the tutorial format (#532)
neo4j/docs-getting-started: modules/ROOT/pages/cypher/recommendation-engine.adoc

= Tutorial: Build a Cypher recommendation engine
:tags: cypher, queries, recommendations, tutorial, filtering, paths
:description: Learn how to build a recommendation engine with Cypher.
:page-aliases: ROOT:guide-build-a-recommendation-engine.adoc

This tutorial uses the xref:appendix/example-data.adoc[Movie dataset] to show how to build two recommendation engines:

* Who an actor hasn't worked with and should work next
* Who can introduce the actors to each other

If you are completely new to Cypher, see the tutorial xref:cypher/intro-tutorial.adoc[Get started with Cypher] to learn how to run basic queries against the same example dataset used here.

== Pre-requisites

After you create a link:https://neo4j.com/product/auradb/?ref=docs-nav-get-started[free Aura instance], use the "Connect" button and select "Query".

> In the Cypher editor, copy and paste the following Cypher and execute the query:

Reviewer: ```suggestion
Copy and paste the following Cypher to the Cypher editor and execute the query:
```

nitpicky, but it's not actually copied from the cypher editor

MERGE (JessicaThompson:Person {name:'Jessica Thompson'})
MERGE (JamesThompson:Person {name:'James Thompson'})

MERGE (JamesThompson)-[:FOLLOWS]->(JessicaThompson)
MERGE (AngelaScope)-[:FOLLOWS]->(JessicaThompson)
MERGE (PaulBlythe)-[:FOLLOWS]->(AngelaScope)

MERGE (JessicaThompson)-[:REVIEWED {summary:'An amazing journey', rating:95}]->(CloudAtlas)
MERGE (JessicaThompson)-[:REVIEWED {summary:'Silly, but fun', rating:65}]->(TheReplacements)
MERGE (JamesThompson)-[:REVIEWED {summary:'The coolest football movie ever', rating:100}]->(TheReplacements)
MERGE (AngelaScope)-[:REVIEWED {summary:'Pretty funny at times', rating:62}]->(TheReplacements)
MERGE (JessicaThompson)-[:REVIEWED {summary:'Dark, but compelling', rating:85}]->(Unforgiven)
MERGE (JessicaThompson)-[:REVIEWED {summary:"Slapstick redeemed only by the Robin Williams and Gene Hackman's stellar performances", rating:45}]->(TheBirdcage)
MERGE (JessicaThompson)-[:REVIEWED {summary:'A solid romp', rating:68}]->(TheDaVinciCode)
MERGE (JamesThompson)-[:REVIEWED {summary:'Fun, but a little far fetched', rating:65}]->(TheDaVinciCode)
MERGE (JessicaThompson)-[:REVIEWED {summary:'You had me at Jerry', rating:92}]->(JerryMaguire);
--

> You should now have a graph with 171 nodes and 253 relationships.

Reviewer: ```suggestion
You now have a graph with 171 nodes and 253 relationships.
```
confidence! :D

== Recommendations

The Movie dataset includes information about actors, directors, and movies they have directed and/or acted in.

> The type of recommendation you can retrieve is based not only on the information that exists in the graph (e.g., what year someone was born), but also how each datapoint interact with each other (e.g., what movies have actors that were only born after 1900).

Reviewer: ```suggestion
The type of recommendation you can retrieve is based not only on the information that exists in the graph (e.g., what year someone was born), but also how each data point interacts with the others (e.g., what movies have actors that were only born after 1900).
```

> Each degree of complexity of a recommendation requires the query to be modified accordingly.

Reviewer: this implies that there is just one query involved in making a recommendation. which might be the case - but maybe we could say "the query or queries involved"

Author: ```suggestion
Each degree of a recommendation's complexity requires the equivalent query to be modified accordingly.
```

Reviewer: maybe "respective"? 
my suggestion was more about "there could be multiple queries involved" which may hold in general, but not for the examples included in this tutorial, so i think i'm fine with the original

Before building the two recommendation engines, you can try simple filtering queries to explore the graph.

=== Simple filtering

> If you want to watch only movies in which Keanu Reeves has acted in, you can find the information with this query:

Reviewer: ```suggestion
If you want to watch only movies which Keanu Reeves has acted in, you can find them with this query:
```

[source,cypher]
--
MATCH (p:Person {name:"Keanu Reeves"})-[:ACTED_IN]->(m:Movie)
RETURN m as Movies
--

The result is:

[role="queryresult",options="header,footer",cols="1*<m"]
|===
| Movies

| "The Matrix"
| "The Matrix Reloaded"
| "The Matrix Revolutions"
| "The Devil's Advocate"
| "The Replacements"
| "Johny Mnemonic"
| "Something's Gotta Give"

1+d|Rows: 7

> |===

Reviewer: do cypher tests apply here? (general question)

The premise of this simple filtering example is using specific information that can be found in properties.
You can apply the same logic to find, for example, in which movie Keanu Reeves played the role "Julian Mercer" by adding the property `role` to the relationship `:ACTED_IN` in the query:

[source,cypher]
--
MATCH (p:Person {name:"Keanu Reeves"})-[:ACTED_IN {roles:['Julian Mercer']}]->(m:Movie)
RETURN m as Movies

> --

Reviewer: if so (cypher tests), then there's no result for this query (test-skip role?)

> And you get the movie "Something's Gotta Give" as the result.

Reviewer: ```suggestion
You get the movie "Something's Gotta Give" as the result.
```

> However, you can get deeper recommendations from the same dataset by applying more complex filters that address multiple connections between the datapoints.

Reviewer: ```suggestion
However, you can get more elaborate recommendations from the same dataset by applying more complex filters that address multiple connections between the data points.
```

"fine-grained"? "precise"? "customized"?

also, two things about "data points":
 * the browser spellchecker suggests the space
 * aren't these just nodes? if so, should we use "nodes" instead? (more occurrences above)

Author: ```suggestion
However, you can get more in-depth recommendations from the same dataset by applying more complex filters that capture multiple connections between the datapoints.
```

=== Finding second-degree co-actors

> One way to take the previous recommendation query to the next level is by trying to find out what actors have worked with co-actors that worked with Keanu Reeves in his movies (i.e., second-degree actors in Keanu's network).

Reviewer: ```suggestion
One way to take the previous recommendation query to the next level is by trying to find out what actors have worked with co-actors that worked with Keanu Reeves in his movies (i.e., second-degree co-actors in Keanu's network).
```

> This query shows all actors Keanu may not have worked with yet and it makes sure not to return anyone who hasn't directly acted with him:

Author: ```suggestion
This query shows all actors Keanu hasn't worked with yet, and it makes sure not to return anyone who hasn't directly acted with him:
```

> This query shows all actors Keanu may not have worked with yet and it makes sure not to return anyone who hasn't directly acted with him:

Reviewer: why the "may" here? based on the data set, we can predict that, right? (maybe not based on reality - but then that's worth pointing out i think)

[source,cypher]
--
MATCH (keanu:Person {name:"Keanu Reeves"})-[:ACTED_IN]->(movie1:Movie)<-[:ACTED_IN]-(coActor:Person)-[:ACTED_IN]->(movie2:Movie)<-[:ACTED_IN]-(coCoActor:Person)
WHERE keanu <> coCoActor
AND NOT (keanu)-[:ACTED_IN]->(:Movie)<-[:ACTED_IN]-(coCoActor)
RETURN coCoActor.name
--

The result is:

[role="queryresult",options="header,footer",cols="1*<m"]
|===
| coCoActor.name

| "Tom Hanks"
| "Halle Berry"
| "Jim Broadbent"
| "Natalie Portman"
| "Stephen Rea"
| "John Hurt"
| "Ben Miles"
| "Tom Hanks"
| "Halle Berry"
| "Jim Broadbent"
| "Natalie Portman"
| "Stephen Rea"
| "John Hurt"
| "Ben Miles"
| "Tom Hanks"
| "Halle Berry"
| "Jim Broadbent"
| "Natalie Portman"
| "Stephen Rea"
| "John Hurt"
| "Ben Miles"

1+d|Rows: 21
|===

Note that a few names appear multiple times.
This is because there are multiple paths to follow from Keanu Reeves to these actors.

> To see which co-co-actors appear most often in Keanu's network, you can take the frequency of occurrences into account by counting the number of paths between Tom Hanks and each `coCoActor` and ordering them by highest to lowest value:

Author: Maybe...?

```suggestion
To see which second-degree co-actors appear most often in Keanu's network, you can take the frequency of occurrences into account by counting the number of paths between Tom Hanks and each `2dCoActor` and ordering them by highest to lowest value:
```

Reviewer: hm.... thing is, the query is not bout Tom Hanks only
it's about all 2nd degree co-actors of Keanu, so Tom Hanks should be replaced with co-actors
`2dCoActor` (in backticks, which usually refers to entities in code listings) doesn't exist in the query either. to make it readable, i think it would be good to  use `coCoActor` for clear reference

Author: I mean, if we change `coCoActor` to `2dActor` then the query would become:

```
MATCH (keanu:Person {name:"Keanu Reeves"})-[:ACTED_IN]->(movie1:Movie)<-[:ACTED_IN]-(coActor:Person)-[:ACTED_IN]->(movie2:Movie)<-[:ACTED_IN]-(2dCoActor:Person)
WHERE keanu <> 2dCoActor
AND NOT (keanu)-[:ACTED_IN]->(:Movie)<-[:ACTED_IN]-(2dCoActor)
RETURN 2dCoActor.name, count(2dCoActor) as frequency
ORDER BY frequency DESC
LIMIT 5
```

You think it looks weird?

Reviewer: it's a variable, so we can select a name freely :) i don't think `2dCoActor` is weird -
if we change it in the query in line 779+ though, we should also update the query in 737+

> To see which second-degree co-actors appear most often in Keanu's network, you can take the frequency of occurrences into account by counting the number of paths between Tom Hanks and each `2dCoActor` and ordering them by highest to lowest value:

Reviewer: Tom Hanks is still wrong, isn't it?
should it be Keanu?

> To see which second-degree co-actors appear most often in Keanu's network, you can take the frequency of occurrences into account by counting the number of paths between Tom Hanks and each `2dCoActor` and ordering them by highest to lowest value:

Author: ```suggestion
To see which second-degree co-actors appear most often in Keanu's network, you can take the frequency of occurrences into account by counting the number of paths between Keanu Reeves and each `2dCoActor` and ordering them by highest to lowest value:
```

> To see which co-co-actors appear most often in Keanu's network, you can take the frequency of occurrences into account by counting the number of paths between Tom Hanks and each `coCoActor` and ordering them by highest to lowest value:

Reviewer: ```suggestion
To see which co-co-actors appear most often in Keanu's network, you can take the frequency of occurrences into account by counting the number of paths between each `coActor` and their `coCoActor`s and ordering them by highest to lowest value:
```

the plural s after `coCoActor` is a crime 🤔 any suggestion?

[source,cypher]
----
MATCH (keanu:Person {name:"Keanu Reeves"})-[:ACTED_IN]->(movie1:Movie)<-[:ACTED_IN]-(coActor:Person)-[:ACTED_IN]->(movie2:Movie)<-[:ACTED_IN]-(coCoActor:Person)
WHERE keanu <> coCoActor
AND NOT (keanu)-[:ACTED_IN]->(:Movie)<-[:ACTED_IN]-(coCoActor)
RETURN coCoActor.name, count(coCoActor) as frequency
ORDER BY frequency DESC
LIMIT 5
----

The result is:

[role="queryresult",options="header,footer",cols="2*<m"]
|===
| coCoActor.name
| frequency

| "Tom Hanks"
| 4

| "Jim Broadbent"
| 3

| "Natalie Portman"
| 3

| "Stephen Rea"
| 3

| "Halle Berry"
| 3

2+d|Rows: 5
|===

One of those "co-co-actors" is Tom Hanks.

=== Exploring the paths between actors and movies

> After the last query, you can now fine-tune it to see which movies and actors are between Keanu and Tom and find out who can introduce them:

Reviewer: ```suggestion

=== Exploring the paths between actors and movies

One of Keanu Reeves' "co-co-actors" is Tom Hanks.
You can fine-tune the last query to see which movies and actors are between Keanu and Tom and find out who can introduce them:
```

[source,cypher]
--
MATCH (keanu:Person {name:"Keanu Reeves"})-[r:ACTED_IN]->(movie1:Movie)<-[r2:ACTED_IN]-(coActor:Person)-[r3:ACTED_IN]->(movie2:Movie)<-[r4:ACTED_IN]-(tom:Person {name:"Tom Hanks"})
WHERE NOT (keanu)-[r]->(:Movie)<-[r2]-(tom)
RETURN keanu, movie1, coActor, movie2, tom, r, r2, r3, r4
--

> The graph result is:

Reviewer: same here, not a result in the technical (testy) sense

image::cocoactors.svg[Graph results of querying showing which movies and actors are between Keanu Reeves and Tom Hank, role="popup-link",width=600]

Now you know that the actors Hugo Weaving and Charlize Theron could be the ones to connect Keanu Reeves and Tom Hanks, as they have worked with both of them in different movies.

> A similar query was shown in the tutorial xref:cypher/intro-tutorial.adoc#_movies_and_actors_connected_to_kevin_bacon[Get started with Cypher], featuring the "Six degrees of Kevin Bacon" game.

Reviewer: ```suggestion
A similar query is shown in the tutorial xref:cypher/intro-tutorial.adoc#_movies_and_actors_connected_to_kevin_bacon[Get started with Cypher], featuring the "Six degrees of Kevin Bacon" game.
```

In both cases, you can see in the graph how many hops exist between people.

== Conclusion

With these Cypher statements, you were able to query actors and movies from the database, and build recommendation engines that show:

> * Who an actor hasn't worked with yet and should do next

Author: ```suggestion
* Who an actor hasn't worked with yet and could do next
```

> * Who an actor hasn't worked with yet and should do next

Reviewer: "should" is a bit suggestive, maybe leave that out
